core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -745,16 +709,54 @@ private[zio] final class FiberContext[E, A](
     }
   }
 
-  private[this] final def exitSupervision: UIO[Any] = ZIO.effectTotal(supervised.pop())
+  private[this] final def daemonEnter(flag: DaemonStatus): UIO[Unit] = ZIO.effectTotal {
+    daemonStatus.push(flag.toBoolean)
+  }
+
+  private[this] final def daemonExit: UIO[Unit] = ZIO.effectTotal {
+    val _ = daemonStatus.popOrElse(false)
+  }
+
+  final def propagateAncestorInterruption(): Unit = {
+    var fiber = self: FiberContext[_, _]
+
+    while (fiber ne null) {
+      addInterruptor(fiber.state.get.interrupted)
+
+      fiber = fiber.parentFiber
+    }
+  }
 
   @inline
-  private[this] final def interruptible: Boolean = interruptStatus.peekOrElse(true)
+  private final def isInterrupted(): Boolean = {
+    propagateAncestorInterruption()
```

Comment:
Question: Why is this method (which is only called from `shouldInterrupt`) doing the propagation logic?

Reply:
I'm going to fix that soon. It is costly so it cannot be done for every iteration of the eval loop (like it is now).

Reply:
Ok 👍